### ISSUE TITLE
Enhancement for certificate relted code

### ISSUE THREAD

**Issue by WillardHu** [kubeedge/kubeedge#5498]
**What would you like to be added/modified**:

- [ ] Encapsulate Token operations.
- [ ] Abstract some methods for certificates, and implement x509 and K8s csr first(implement gmx509 in the future).
- [ ] Abstract some methods for PKI(tls.Config), and implement standard TLS(implement GMTLS in the future).
- [ ] Abstract some mehtods for Server, and implement websocket, quic and webservice(implement gmwebsocket, gmquic and gmwebservice in the future).
- [ ] Abstract some methods for HubClient, and implement websocket and quic(implement gmwebsocket and gmquic in the future).
- [ ] Abstract some methods for TunnelClient, and implement webservice(implement gmwebservice in the future).

**Why is this needed**:

1. We maintain the functions of a business in one place to improve readability and maintainability.
2. We should use a unique CA to sign the server certificate whenever possible.
3. We abstract the certificate, PKI, Server and Client to extend GMTLS in the future(this issue is the pre-task of the #5334).

**Comment by vaidikcode** [comment]
@WillardHu 
for 3rd step
 **Encapsulate Token Operations**: 
   - I'll define a `TokenManager` interface to centralize token management tasks. This will include methods for generating, validating, and refreshing tokens. The goal is to ensure token management is reusable, easy to maintain, and independent of the rest of the logic.

 **Abstract Certificate Management**: 
   - A `CertificateManager` interface will be introduced to handle certificate operations. Initially, this will focus on x509 certificates (loading, validating, and renewing). This allows for centralized certificate management, ensuring that certificates are consistently handled across the system. Later, we will extend it to support GMX509 for regional compliance.

**Abstract PKI Methods for `tls.Config`**:
   - A `PKIManager` interface will be created to manage TLS configurations (`tls.Config`). This will include methods to generate the TLS configuration, load certificates, add trusted CAs, and validate certificates. The initial implementation will support standard TLS, with future support for GMTLS (a cryptographic standard) to follow. This abstraction will ensure that the TLS setup is reusable and consistent, while also being flexible enough to incorporate future standards.

**Integrate with Server and Client**: 
   - The `PKIManager` and `CertificateManager` interfaces will be integrated into the server and client code. This will simplify TLS configuration across the application and ensure that the TLS setup is uniform. The abstractions will also make it easier to extend support for other protocols in the future, such as GMWebSocket, GMQUIC, and GMWebservice.

 **Testing and Validation**:
   - Unit tests will be written for each method within the `PKIManager` and `CertificateManager` interfaces to verify correctness.
   

Please let me know if there are any additional methods or other functionality that should be considered for this PR. I want to ensure I’m covering everything necessary, or if there are any other concerns or improvements before proceeding.

**Comment by vaidikcode** [comment]
@WillardHu 
Can I also work on this?
for 3rd step-

 **Abstract PKI Methods for `tls.Config`**:
   - I'll introduce a `PKIManager` interface for handling TLS configuration. This interface will include methods for generating a `tls.Config`, loading certificates, adding trusted CAs, and validating certificates.
   - The initial implementation will support standard TLS, ensuring that all operations are abstracted and reusable. Later, we can extend this to support GMTLS for additional cryptographic standards.

 **Integrating PKI Abstractions with Server and Client**:
   - The `PKIManager` will be integrated with server and client components to centralize TLS configuration and certificate management. This will simplify the configuration process, ensuring consistency across different parts of the system.
   - By abstracting TLS operations, we can easily extend support for new protocols in the future (e.g., GMWebSocket, GMQUIC, and GMWebservice).

 **Testing and Validation**:
   - Unit tests will be written for each method in the `PKIManager` interface, focusing on generating and validating TLS configurations.
   - Integration tests will be added to ensure that the server and client configurations are working as expected with the new abstractions.

Please let me know if there are any additional methods or functionality I should consider for this step or any concerns before proceeding.

**Comment by WillardHu** [comment]
Thank you for your interest in this feature.

The native golang tls package does not support extended GMTLS. You can try to investigate whether you can do something with reflection. If not, all C/S tls handshake codes must be rewritten, which will be a lot of work and much less cost-effective than using a proxy server.

**Comment by vaidikcode** [comment]
I’ve been investigating using reflection to inject GMTLS support into the crypto/tls package, by accessing and modifying unexported fields to replace standard cryptographic operations (e.g., SM2, SM3, SM4 for certificates and handshake). However, reflection introduces risks such as brittleness with future Go updates and security/performance concerns. As a more stable approach, I plan to focus on task 3, which abstracts PKI operations using tls.Config, so that GMTLS support can be more easily integrated later. I’ll open a draft PR for this now and gradually move toward GMTLS integration in future. wdyt?

**Comment by WillardHu** [comment]
The key problem is the server and client code, they have no way to use abstraction. 

- HTTP Server
  ```golang
  package http

  import "crypto/tls"
  
  type Server struct {
      TLSConfig *tls.Config
  }
  ```
- QUIC Server (github.com/lucas-clemente/quic-go)
  ```golang
  import "crypto/tls"
  func ListenAddr(addr string, tlsConf *tls.Config, config *Config) (Listener, error) {...}
  ```
- WebSocket Server using http server

This means that we will need to modify them to truly support GMTLS in the future. 😂

**Comment by vaidikcode** [comment]
I understand the architectural constraint you are highlighting. This will require modifying the core server implementation, and there's no way around it for now.😅

**Comment by WillardHu** [comment]
> I understand the architectural constraint you are highlighting. This will require modifying the core server implementation, and there's no way around it for now.😅

Yes, so the simplest solution to use GMTLS is to use a proxy service.

```mermaid
flowchart LR
  subgraph EdgeNode
    A[EdgeCore] --TLS--> B[Proxy Service]
  end
  subgraph Cloud
    B --GMTLS--> C[Proxy Service]
    C --TLS--> D[CloudCore]
  end
```
